README.txt
1. Create a BE User "_cli_dkd_news2direct", which have the permission to select and modify tables of direct_mail, tt_news and your newsletter recipients table (tt_address or fe_user)

2. after installing the dkd_news2directmail extension, you'll find a new module (web > News 2 Directmail). Upon opening this module you'll have to choose a SysFolder containing your directmail.

3. Next you can put in some configuration parameters:
	- recipient list -> you can only put ONE recipient ID, which you have created with directmail extension (see Directmail manual how to do this)
	- time interval -> is the time interval between the news2directmail cli script compiling the news and directmail cli script start sending the newsletter
	- Domain of internal link -> because the base URL of the link is not known in the cli mode, you have to set it here.
	
4. Next you have to call the cli script from the command line. the command is:
	
	/absolute_path_to_your_site/typo3conf/ext/dkd_news2directmail/cli/cli_dkdnews2directmail.php xxx
	
	where xxx is the page ID, which contains tt_news extension displaying the news detail.
	

To test if your page is correctly configured to show tt_news, try to open it with browser
	www.example.com/index.php?id=xxx&tx_ttnews[tt_news]=yyy
where xxx is the page ID and yyy is the tt_news uid.

ChangeLog
10-05-28 Ivan Kartolo <ivan.kartolo@dkd.de>
	* default dmail uid is zero. SQL is updated
	* removed support for older direct_mail < 2.5.0

10-05-26 Ivan Kartolo <ivan.kartolo@dkd.de>
	* replaced deprecated functions

06-11-23 Ivan Kartolo  <ivan.kartolo@dkd.de>

	* Initial release


ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "dkd_news2directmail".
#
# Auto generated 11-11-2010 11:03
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'dkd_news2directmail',
	'description' => 'Automatically sends tt_news as a directmail',
	'category' => 'be',
	'author' => 'Ivan Kartolo',
	'author_email' => 'ivan.kartolo@dkd.de',
	'shy' => '',
	'dependencies' => 'tt_news,direct_mail',
	'conflicts' => '',
	'priority' => '',
	'module' => 'mod1,cli',
	'state' => 'beta',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => 'tt_news',
	'clearCacheOnLoad' => 0,
	'lockType' => '',
	'author_company' => 'd.k.d Internet Service GmbH',
	'version' => '1.0.0',
	'constraints' => array(
		'depends' => array(
			'tt_news' => '',
			'direct_mail' => '2.5.x',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:20:{s:9:"ChangeLog";s:4:"f85f";s:10:"README.txt";s:4:"a4cb";s:12:"ext_icon.gif";s:4:"b831";s:14:"ext_tables.php";s:4:"04a8";s:14:"ext_tables.sql";s:4:"9ff2";s:16:"locallang_db.xml";s:4:"1772";s:37:"cli/class.ext_tx_directmail_dmail.php";s:4:"ce9b";s:30:"cli/cli_dkdnews2directmail.php";s:4:"2f76";s:12:"cli/conf.php";s:4:"7354";s:14:"doc/manual.sxw";s:4:"e427";s:19:"doc/wizard_form.dat";s:4:"c93d";s:20:"doc/wizard_form.html";s:4:"a8fd";s:14:"mod1/clear.gif";s:4:"cc11";s:13:"mod1/conf.php";s:4:"ea2f";s:14:"mod1/index.php";s:4:"eaa3";s:18:"mod1/locallang.xml";s:4:"05a5";s:22:"mod1/locallang_mod.xml";s:4:"dcfe";s:19:"mod1/moduleicon.gif";s:4:"b831";s:45:"static/tt_news_single_plaintext/constants.txt";s:4:"25ca";s:41:"static/tt_news_single_plaintext/setup.txt";s:4:"2bdc";}',
	'suggests' => array(
	),
);

?>

mod1/index.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2006 Ivan Kartolo <ivan.kartolo@dkd.de>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/
/**
 * Module 'News 2 Directmail' for the 'dkd_news2directmail' extension.
 *
 * @author	Ivan Kartolo <ivan.kartolo@dkd.de>
 */



	// DEFAULT initialization of a module [BEGIN]
unset($MCONF);
require ("conf.php");
require ($BACK_PATH."init.php");
require ($BACK_PATH."template.php");
$LANG->includeLLFile("EXT:dkd_news2directmail/mod1/locallang.xml");
require_once (PATH_t3lib."class.t3lib_scbase.php");
$BE_USER->modAccess($MCONF,1);	// This checks permissions and exits if the users has no permission for entry.
	// DEFAULT initialization of a module [END]

class tx_dkdnews2directmail_module1 extends t3lib_SCbase {
	var $pageinfo;
	var $TSconfPrefix = 'mod.web_modules.news2directmail.';
	
	/**
	 * Initializes the Module
	 * @return	void
	 */
	function init()	{
		global $BE_USER,$LANG,$BACK_PATH,$TCA_DESCR,$TCA,$CLIENT,$TYPO3_CONF_VARS;

		parent::init();
		
		$this->modList = $this->getListOfBackendModules(array('dmail'),$this->perms_clause,$BACK_PATH);
		$temp = t3lib_BEfunc::getModTSconfig($this->id,'mod.web_modules.news2directmail');
		$this->params = $temp['properties'];
		$this->implodedParams = t3lib_BEfunc::implodeTSParams($this->params);
		
		$this->updatePageTS();
		/*
		if (t3lib_div::_GP("clear_all_cache"))	{
			$this->include_once[]=PATH_t3lib."class.t3lib_tcemain.php";
		}
		*/
	}

	/**
	 * Main function of the module. Write the content to $this->content
	 * If you chose "web" as main module, you will need to consider the $this->id parameter which will contain the uid-number of the page clicked in the page tree
	 *
	 * @return	[type]		...
	 */
	function main()	{
		global $BE_USER,$LANG,$BACK_PATH,$TCA_DESCR,$TCA,$CLIENT,$TYPO3_CONF_VARS;

		// Access check!
		// The page will show only if there is a valid page and if this page may be viewed by the user
		$this->pageinfo = t3lib_BEfunc::readPageAccess($this->id,$this->perms_clause);
		$access = is_array($this->pageinfo) ? 1 : 0;

		if (($this->id && $access) || ($BE_USER->user["admin"] && !$this->id))	{

				// Draw the header.
			$this->doc = t3lib_div::makeInstance("mediumDoc");
			$this->doc->backPath = $BACK_PATH;
			$this->doc->form='<form action="" method="POST">';

				// JavaScript
			$this->doc->JScode = '
				<script language="javascript" type="text/javascript">
					script_ended = 0;
					function jumpToUrl(URL)	{
						document.location = URL;
					}
				</script>
			';
			$this->doc->postCode='
				<script language="javascript" type="text/javascript">
					script_ended = 1;
					if (top.fsMod) top.fsMod.recentIds["web"] = 0;
				</script>
			';

			$headerSection = $this->doc->getHeader("pages",$this->pageinfo,$this->pageinfo["_thePath"])."<br />".$LANG->sL("LLL:EXT:lang/locallang_core.xml:labels.path").": ".t3lib_div::fixed_lgd_cs($this->pageinfo["_thePath"],50);

			$this->content.=$this->doc->startPage($LANG->getLL("title"));
			$this->content.=$this->doc->header($LANG->getLL("title"));
			$this->content.=$this->doc->spacer(5);
//			$this->content.=$this->doc->section("",$this->doc->funcMenu($headerSection,t3lib_BEfunc::getFuncMenu($this->id,"SET[function]",$this->MOD_SETTINGS["function"],$this->MOD_MENU["function"])));
//			$this->content.=$this->doc->divider(5);

			$module = $this->pageinfo['module'];

			if (!$module)	{
				$pidrec=t3lib_BEfunc::getRecord('pages',intval($this->pageinfo['pid']));
				$module=$pidrec['module'];
			}
			if ($module == 'dmail') {
				$this->content.=$this->doc->section('',$this->doc->funcMenu($headerSection, t3lib_BEfunc::getFuncMenu($this->id,'SET[dmail_mode]',$this->MOD_SETTINGS['dmail_mode'],$this->MOD_MENU['dmail_mode']).t3lib_BEfunc::cshItem($this->cshTable,'',$BACK_PATH)));

					// Render content:
				$this->moduleContent();
			} else {
				$this->content.=$this->doc->section($LANG->getLL('dmail_folders').t3lib_BEfunc::cshItem($this->cshTable,'folders',$BACK_PATH), $this->modList['list'], 1, 1, 0 , TRUE);
			}


			// ShortCut
			if ($BE_USER->mayMakeShortcut())	{
				$this->content.=$this->doc->spacer(20).$this->doc->section("",$this->doc->makeShortcutIcon("id",implode(",",array_keys($this->MOD_MENU)),$this->MCONF["name"]));
			}

			$this->content.=$this->doc->spacer(10);
		} else {
				// If no access or if ID == zero

			$this->doc = t3lib_div::makeInstance("mediumDoc");
			$this->doc->backPath = $BACK_PATH;

			$this->content.=$this->doc->startPage($LANG->getLL("title"));
			$this->content.=$this->doc->header($LANG->getLL("title"));
			$this->content.=$this->doc->spacer(5);
			$this->content.=$this->doc->spacer(10);
		}
	}

	/**
	 * Prints out the module HTML
	 *
	 * @return	void
	 */
	function printContent()	{

		$this->content.=$this->doc->endPage();
		echo $this->content;
	}

	/**
	 * Generates the module content
	 *
	 * @return	void
	 */
	function moduleContent()	{
		$this->content .= $this->cmd_conf();
		$this->updatePageTS();
	}


	function updatePageTS()	{
		global $BE_USER;

		if ($BE_USER->doesUserHaveAccess(t3lib_BEfunc::getRecord( 'pages', $this->id), 2)) {
			$pageTS = t3lib_div::_GP('pageTS');
			if (is_array($pageTS))	{
				t3lib_BEfunc::updatePagesTSconfig($this->id,$pageTS,$this->TSconfPrefix);
				header('Location: '.t3lib_div::locationHeaderUrl(t3lib_div::getIndpEnv('REQUEST_URI')));
			}
		}
	}
	
	function cmd_conf() {
		global $TYPO3_DB, $LANG;

		$configArray = array(
			'spacer0' => $LANG->getLL('configure_news2directmail'),
			'recipient_list' => array('short', $LANG->getLL('recipient_list'), $LANG->getLL('recipient_list.description').'<br />'.$LANG->getLL('recipient_list.details')),
			'send_time' => array('short', $LANG->getLL('send_time'), $LANG->getLL('send_time.description').'<br />'.$LANG->getLL('send_time.details')),
			'URL' => array('string', $LANG->getLL('URL'), $LANG->getLL('URL.description').'<br />'.$LANG->getLL('URL.details')),
			);

		$theOutput.= $this->doc->section($LANG->getLL('configure_direct_mail_module'),str_replace('Update configuration', $LANG->getLL('configure_update_configuration'), t3lib_BEfunc::makeConfigForm($configArray,$this->implodedParams,'pageTS')),1,1,0, TRUE);
		return $theOutput;
	}
	
	/**
	 * Returns "list of backend modules". Most likely this will be obsolete soon / removed. Don't use.
	 * Usage: 0
	 *
	 * @param	array		Module names in array. Must be "addslashes()"ed
	 * @param	string		Perms clause for SQL query
	 * @param	string		Backpath
	 * @param	string		The URL/script to jump to (used in A tag)
	 * @return	array		Two keys, rows and list
	 * @internal
	 * @deprecated since TYPO3 3.6, this function will be removed in TYPO3 4.5.
	 * @obsolete
	 */
	public static function getListOfBackendModules($name, $perms_clause, $backPath = '', $script = 'index.php') {
		$res = $GLOBALS['TYPO3_DB']->exec_SELECTquery('*', 'pages', 'doktype!=255 AND module IN (\''.implode('\',\'', $name).'\') AND'.$perms_clause.t3lib_BEfunc::deleteClause('pages'));
		if (!$GLOBALS['TYPO3_DB']->sql_num_rows($res))	return false;

		$out = '';
		$theRows = array();
		while($row = $GLOBALS['TYPO3_DB']->sql_fetch_assoc($res)) {
			$theRows[] = $row;
			$out.='<span class="nobr"><a href="'.htmlspecialchars($script.'?id='.$row['uid']).'">'.
					t3lib_iconWorks::getIconImage('pages', $row, $backPath, 'title="'.htmlspecialchars(t3lib_BEfunc::getRecordPath($row['uid'], $perms_clause, 20)).'" align="top"').
					htmlspecialchars($row['title']).
					'</a></span><br />';
		}
		$GLOBALS['TYPO3_DB']->sql_free_result($res);

		return array('rows'=>$theRows, 'list'=>$out);
	}
	
	
}



if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/dkd_news2directmail/mod1/index.php'])	{
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/dkd_news2directmail/mod1/index.php']);
}




// Make instance:
$SOBE = t3lib_div::makeInstance('tx_dkdnews2directmail_module1');
$SOBE->init();

// Include files?
foreach($SOBE->include_once as $INC_FILE)	include_once($INC_FILE);

$SOBE->main();
$SOBE->printContent();

?>

cli/cli_dkdnews2directmail.php
<?php
/* cli_dkdnews2directmail.php,v 1.1 2006/11/23 17:17:40 dkd-kartolo Exp */
	// Defining circumstances for CLI mode:
define('TYPO3_cliMode', TRUE);


	// Defining PATH_thisScript here: Must be the ABSOLUTE path of this script in the right context:
	// This will work as long as the script is called by it's absolute path!
// define('PATH_thisScript',$_ENV['_']?$_ENV['_']:$_SERVER['_']);
define('PATH_thisScript',$_SERVER['argv'][0]);


	// Change working directory to the directory of the script.
chdir( dirname( PATH_thisScript ) );


	// Include configuration file:
require(dirname(PATH_thisScript).'/conf.php');

	// Include init file:
require(dirname(PATH_thisScript).'/'.$BACK_PATH.'init.php');

 // Define path to TSlib
define('PATH_tslib', PATH_site.'tslib/');


// HERE you run your application!
define( 'TYPO3_MODE', 'BE' );

require_once(PATH_t3lib.'class.t3lib_div.php');
require_once(PATH_t3lib.'class.t3lib_extmgm.php');
require_once(PATH_t3lib.'class.t3lib_cs.php');

require_once(PATH_t3lib."config_default.php");
if (!defined ("TYPO3_db")) 	die ("The configuration file was not included.");


// Require DB classes
require_once(PATH_t3lib.'class.t3lib_db.php');	// The database library
$TYPO3_DB = t3lib_div::makeInstance('t3lib_DB');


// Set debug options for DB
// $GLOBALS['TYPO3_DB']->debugOutput = true;
// $GLOBALS['TYPO3_DB']->store_lastBuiltQuery = true;


// Connect to the database, needed at least for t3lib_extmgm to work
$result = $GLOBALS['TYPO3_DB']->sql_pconnect(TYPO3_db_host, TYPO3_db_username, TYPO3_db_password);
if (!$result)	{
	die("Couldn't connect to database at ".TYPO3_db_host);
}
$GLOBALS['TYPO3_DB']->sql_select_db(TYPO3_db);

// Application code starts here


// > 2.5.x is installed
require(dirname(PATH_thisScript).'/class.ext_tx_directmail_dmail.php');
// Make instance:
$SOBE = t3lib_div::makeInstance('ext_tx_directmail_dmail');

// argv[1] => newsletter uid
$SOBE->init($argv[1]);

// Include files?
foreach($SOBE->include_once as $INC_FILE)	include_once($INC_FILE);

$SOBE->main();

?>


cli/class.ext_tx_directmail_dmail.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2006 dkd-ivan
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/*
 * class.ext_mod_web_dmail.php,v 1.1 2006/11/23 17:17:40 dkd-kartolo Exp
 */
require_once(t3lib_extMgm::extPath('direct_mail').'mod2/class.tx_directmail_dmail.php');
require_once (PATH_t3lib.'class.t3lib_tstemplate.php');

class ext_tx_directmail_dmail extends tx_directmail_dmail {
	
	/*
	 * what need to be initialized??
	 * 
	 * @param	int		page id of the newsletter
	 */
//TODO: init from mod_web_dmail? what else need to be initialized?
	function init($newsletter_uid){
		global $LANG,$BACK_PATH,$TCA_DESCR,$TCA,$CLIENT,$TYPO3_CONF_VARS, $TYPO3_DB, $MCONF;
		
		$this->include_once[]=PATH_t3lib.'class.t3lib_tcemain.php';
		
			// initialize the TS template
		$GLOBALS['TT'] = new t3lib_timeTrack;
		$this->tmpl = t3lib_div::makeInstance('t3lib_TStemplate');
		$this->tmpl->init();
		
			// initialize the page selector
		$this->sys_page = t3lib_div::makeInstance('t3lib_pageSelect');
		$this->sys_page->init(true);
		
		//get the id of dmail sysfolder
		$this->get_pid($newsletter_uid);
		
		//PageTS for dmail
		$temp = t3lib_BEfunc::getModTSconfig($this->id,'mod.web_modules.dmail');
		$this->params = $temp['properties'];
		
		//PageTS for news2dmail
		unset($temp);
		$temp = t3lib_BEfunc::getModTSconfig($this->id,'mod.web_modules.news2directmail');
		$this->params_news2dmail = $temp['properties'];
		
		t3lib_div::loadTCA('sys_dmail');

	}
	
	/*
	 * 
	 */
//TODO: how to render plaintext ?  
	function main(){
		global $TYPO3_DB;
		
		//get tt_news_uid
		$this->tt_news = $this->get_ttnews_uid();
		if(!empty($this->tt_news['uid'])){
			$this->perms_clause = $GLOBALS['BE_USER']->getPagePermsClause(1);
			
			$this->pageinfo = t3lib_BEfunc::readPageAccess($this->id,$this->perms_clause);
	
				//set param for cmd_send_mail
			$this->mailgroup_uid = $this->params_news2dmail['recipient_list'];
			$this->mailingMode_mailGroup = 1;
			$this->send_mail_datetime = $GLOBALS['EXEC_TIME'] + ($this->params_news2dmail['send_time'] * 60);
			$this->CMD = 'send_mail_final';
			
			//Set HTML und plain param
			$this->params['HTMLParams'] = '&no_cache=1&tx_ttnews[tt_news]='.$this->tt_news['uid'];
			$this->params['plainParams'] = '&no_cache=1&tx_ttnews[tt_news]='.$this->tt_news['uid'].'&type=99';
			
			//create dmail record
			$this->createDMail();
			
			// auto fetching
			// Read new record (necessary because TCEmain sets default field values)
	  		$dmailRec = t3lib_BEfunc::getRecord ('sys_dmail',$this->sys_dmail_uid);

	  		// Set up URLs from record data for fetch command
	  		$this->setURLs($dmailRec);
			$this->cmd_fetch($dmailRec,TRUE);
	
	  		//dmail record is ready to send
	  		$this->cmd_send_mail($dmailRec);
	  		
	  		//TODO: put dmail record uid to news
	  		$TYPO3_DB->exec_UPDATEquery(
	  			'tt_news',
	  			'uid = '.$this->tt_news['uid'],
	  			array(
	  				'tx_dkdnews2directmail_dmail_uid' => $this->sys_dmail_uid
	  			)
	  		);
		}
	}
	
		/**
	 * Creates a directmail entry in th DB.
	 * @return	[type]		...
	 */
	function createDMail()	{
		global $TCA, $TYPO3_CONF_VARS;

			//only from internal uid
		$createMailFrom_UID = $this->newsletter_uid;	// Internal page

		if ($createMailFrom_UID)	{
				// Set default values:
			$dmail = array();
			$dmail['sys_dmail']['NEW'] = array (
				'from_email'		=> $this->params['from_email'],
				'from_name'		=> $this->params['from_name'],
				'replyto_email'		=> $this->params['replyto_email'],
				'replyto_name'		=> $this->params['replyto_name'],
				'return_path'		=> $this->params['return_path'],
				'priority'		=> $this->params['priority'],
				'use_domain'		=> $this->params['use_domain'],
				'use_rdct'		=> $this->params['use_rdct'],
				'long_link_mode'	=> $this->params['long_link_mode'],
				'organisation'		=> $this->params['organisation'],
				'authcode_fieldList'	=> $this->params['authcode_fieldList']
				);

			$dmail['sys_dmail']['NEW']['sendOptions'] = $TCA['sys_dmail']['columns']['sendOptions']['config']['default'];
			$dmail['sys_dmail']['NEW']['long_link_rdct_url'] = $this->getUrlBase($this->params['use_domain']);

				// If params set, set default values:
			if (isset($this->params['sendOptions']))	$dmail['sys_dmail']['NEW']['sendOptions'] = $this->params['sendOptions'];
			if (isset($this->params['includeMedia'])) 	$dmail['sys_dmail']['NEW']['includeMedia'] = $this->params['includeMedia'];
			if (isset($this->params['flowedFormat'])) 	$dmail['sys_dmail']['NEW']['flowedFormat'] = $this->params['flowedFormat'];
			if (isset($this->params['HTMLParams']))		$dmail['sys_dmail']['NEW']['HTMLParams'] = $this->params['HTMLParams'];
			if (isset($this->params['plainParams']))	$dmail['sys_dmail']['NEW']['plainParams'] = $this->params['plainParams'];
			if (isset($this->params['direct_mail_encoding']))	$dmail['sys_dmail']['NEW']['encoding'] = $this->params['direct_mail_encoding'];

			if (t3lib_div::testInt($createMailFrom_UID))	{
				$createFromMailRec = t3lib_BEfunc::getRecord ('pages',$createMailFrom_UID);
				if (t3lib_div::inList($TYPO3_CONF_VARS['FE']['content_doktypes'],$createFromMailRec['doktype']))	{
//TODO: mail subject is news title? charset?
					$dmail['sys_dmail']['NEW']['subject'] = $this->tt_news['title'];
					$dmail['sys_dmail']['NEW']['type'] = 0;
					$dmail['sys_dmail']['NEW']['page'] = $createFromMailRec['uid'];
					$dmail['sys_dmail']['NEW']['charset'] = $this->getPageCharSet($createFromMailRec['uid']);
					$dmail['sys_dmail']['NEW']['pid'] = $this->pageinfo['uid'];
				}
			} 

			if ($dmail['sys_dmail']['NEW']['pid'] && $dmail['sys_dmail']['NEW']['sendOptions']) {
				$tce = t3lib_div::makeInstance('t3lib_TCEmain');
				$tce->stripslashes_values=0;
				$tce->start($dmail,Array());
				$tce->process_datamap();
				$this->sys_dmail_uid = $tce->substNEWwithIDs['NEW'];
			} else {
				if (!$dmail['sys_dmail']['NEW']['sendOptions']) {
					
					//error??
					$this->error = 'no_valid_url';
					die('no_valid_url');
				}
			}
		}
	}

	/**
	 * [Describe function...]
	 *
	 * @param	[type]		$row: ...
	 * @return	[type]		...
	 */
	function cmd_send_mail($row)	{
		global $LANG, $TYPO3_DB;

			// Preparing mailer
		$htmlmail = t3lib_div::makeInstance('dmailer');
		$htmlmail->start();
		$htmlmail->dmailer_prepare($row);

		$sentFlag=false;

		if ($this->mailingMode_mailGroup && $this->sys_dmail_uid && intval($this->mailgroup_uid))	{
				// Update the record:
			$result = $this->cmd_compileMailGroup(intval($this->mailgroup_uid));
			$query_info=$result['queryInfo'];

			$distributionTime = intval($this->send_mail_datetime);
			$distributionTime = $distributionTime<time() ? time() : $distributionTime;

			$updateFields = array(
				'scheduled' => $distributionTime,
				'query_info' => serialize($query_info)
				);
			$TYPO3_DB->exec_UPDATEquery(
				'sys_dmail',
				'uid='.intval($this->sys_dmail_uid),
				$updateFields
				);

			$sentFlag=true;

			$this->noView=1;
		}
				
			// Setting flags:
		if ($sentFlag && $this->CMD=='send_mail_final')	{
				// Update the record:
			$TYPO3_DB->exec_UPDATEquery(
				'sys_dmail',
				'uid='.intval($this->sys_dmail_uid),
				array('issent' => 1)
				);
		}
//		return $theOutput;
	}

	/**
	 * [Describe function...]
	 *
	 * @param	[type]		$formname: ...
	 * @return	string		urlbase
	 */
	function getUrlBase($domainUid) {
		global $TYPO3_DB;

		$domainName = '';
		$scheme = '';
		$port = '';
		if ($domainUid) {
			$res_domain = $TYPO3_DB->exec_SELECTquery(
				'domainName',
				'sys_domain',
				'uid='.intval($domainUid).
					t3lib_BEfunc::deleteClause('sys_domain')
				);
			if ($row_domain = $TYPO3_DB->sql_fetch_assoc($res_domain)) {
				$domainName = $row_domain['domainName'];
				$url_parts = parse_url(t3lib_div::getIndpEnv('TYPO3_REQUEST_DIR'));
				$scheme = $url_parts['scheme'];
				$port = $url_parts['port'];
			}
		}

		return ($domainName ? (($scheme?$scheme:'http') . '://' . $domainName . ($port?':'.$port:'') . '/') : $this->params_news2dmail['URL'].'/').'index.php';
	} 
	
	/*[Describe function...]
	 *
	 * @param	[type]		$row: ...
	 * @return	[type]		...
	 */
	function cmd_fetch($row)        {
		global $TCA, $TYPO3_DB, $LANG;

		$theOutput = '';
		$errorMsg = '';
		$warningMsg = '';
		$content ='';
		$success = FALSE;

			// Make sure long_link_rdct_url is consistent with use_domain.
		$this->urlbase = $this->getUrlBase($row['use_domain']);
		$row['long_link_rdct_url'] = $this->urlbase;

			// Compile the mail
		$htmlmail = t3lib_div::makeInstance('dmailer');
		if($this->params['enable_jump_url']) {
			$htmlmail->jumperURL_prefix = $this->urlbase.'?id='.$row['page'].'&rid=###SYS_TABLE_NAME###_###USER_uid###&mid=###SYS_MAIL_ID###&aC=###SYS_AUTHCODE###&jumpurl=';
			$htmlmail->jumperURL_useId=1;
		}
		$htmlmail->start();
		$htmlmail->charset = $row['charset'];
		$htmlmail->useBase64();
		$htmlmail->http_username = $this->params['http_username'];
		$htmlmail->http_password = $this->params['http_password'];
		$htmlmail->includeMedia = $row['includeMedia'];

		if ($this->url_plain) {
			$content = t3lib_div::getURL($this->addUserPass($this->url_plain));
				// If we have a conversion table for certain entities and the used charset we do a search and replace.
			if ( is_file( t3lib_extMgm::extPath( 'direct_mail' ) . 'res/char-conv/' . $row['charset'].'.php' ) ) {
				require_once( t3lib_extMgm::extPath( 'direct_mail' ) . 'res/char-conv/'.$row['charset'].'.php' );
				if ( is_array( $charConv ) ) {
					foreach( $charConv AS $cvKey => $cvVal ) {
						$search[] = $cvKey;
						$replace[] = $cvVal;
					}
					$content = str_replace( $search, $replace, $content );
				}
			}
			$htmlmail->addPlain($content);
			if (!$content || !$htmlmail->theParts['plain']['content']) {
				$errorMsg .= 'dmail_no_plain_content'.chr(10);
			} elseif (!strstr(base64_decode($htmlmail->theParts['plain']['content']),'<!--DMAILER_SECTION_BOUNDARY')) {
				$warningMsg .= 'dmail_no_plain_boundaries'.chr(10);
			}
		}
		if ($this->url_html) {
			$success = $htmlmail->addHTML($this->url_html);    // Username and password is added in htmlmail object
			if (!$row['charset']) {		// If no charset was set, we have an external page.
					// Try to auto-detect the charset of the message
				$matches = array();
				$res = preg_match('/<meta[\s]+http-equiv="Content-Type"[\s]+content="text\/html;[\s]+charset=([^"]+)"/m', $htmlmail->theParts['html_content'], $matches);
				if ($res==1) {
					$htmlmail->charset = $matches[1];
				} elseif (isset($this->params['direct_mail_charset'])) {
					$htmlmail->charset = $LANG->csConvObj->parse_charset($this->params['direct_mail_charset']);
				} else {
					$htmlmail->charset = 'iso-8859-1';
				}
				$htmlmail->useBase64();   // Reset content-type headers with new charset
			}
			if ($htmlmail->extractFramesInfo()) {
				$errorMsg .= 'dmail_frames_not allowed'.chr(10);
			} elseif (!$success || !$htmlmail->theParts['html']['content']) {
				$errorMsg .= 'dmail_no_html_content'.chr(10);
			} elseif (!strstr(base64_decode($htmlmail->theParts['html']['content']),'<!--DMAILER_SECTION_BOUNDARY')) {
				$warningMsg .= 'dmail_no_html_boundaries'.chr(10);
			}
		}

		$attachmentArr = t3lib_div::trimExplode(',', $row['attachment'],1);
		if (count($attachmentArr))	{
			t3lib_div::loadTCA('sys_dmail');
			$upath = $TCA['sys_dmail']['columns']['attachment']['config']['uploadfolder'];
			while(list(,$theName)=each($attachmentArr))	{
				$theFile = PATH_site.$upath.'/'.$theName;
				if (@is_file($theFile))	{
					$htmlmail->addAttachment($theFile, $theName);
				}
			}
		}

		if (!$errorMsg) {
				// Update the record:
			$htmlmail->theParts['messageid'] = $htmlmail->messageid;
			$mailContent = serialize($htmlmail->theParts);
			$updateFields = array(
				'issent' => 0,
				'charset' => $htmlmail->charset,
				'mailContent' => $mailContent,
				'renderedSize' => strlen($mailContent),
				'long_link_rdct_url' => $this->urlbase
				);
			$TYPO3_DB->exec_UPDATEquery(
				'sys_dmail',
				'uid='.intval($this->sys_dmail_uid),
				$updateFields
				);

				// Read again:
			$res = $TYPO3_DB->exec_SELECTquery(
				'*',
				'sys_dmail',
				'pid='.intval($this->id).
					' AND uid='.intval($this->sys_dmail_uid).
					t3lib_BEfunc::deleteClause('sys_dmail')
					);
			$row = $TYPO3_DB->sql_fetch_assoc($res);

echo $warningMsg;			
		} else {
			
echo $errorMsg;
			$this->noView = 1;
		}

		return $theOutput;
	}
	
	/**
 	 * Set up URL variables for this $row.
 	 *
	 */
 	function setURLs($row)	{
		// Finding the domain to use
 		$this->urlbase = $this->getUrlBase($row['use_domain']);
 
 			// Finding the url to fetch content from
 		switch((string)$row['type'])	{
 			case 1:
 				$this->url_html = $row['HTMLParams'];
 				$this->url_plain = $row['plainParams'];
 				break;
 			default:
 				$this->url_html = $this->urlbase.'?id='.$row['page'].$row['HTMLParams'];
 				$this->url_plain = $this->urlbase.'?id='.$row['page'].$row['plainParams'];
 				break;
 		}
 
 		if (!($row['sendOptions']&1) || !$this->url_plain)	{	// plain
 			$this->url_plain='';
 		} else {
 			$urlParts = @parse_url($this->url_plain);
 			if (!$urlParts['scheme'])	{
 				$this->url_plain='http://'.$this->url_plain;
 			}
 		}
 		if (!($row['sendOptions']&2) || !$this->url_html)	{	// html
 			$this->url_html='';
 		} else {
 			$urlParts = @parse_url($this->url_html);
			if (!$urlParts['scheme'])	{
 				$this->url_html='http://'.$this->url_html;
 			}
 		}
 	}
  
	/*
	 * @param	int		newsletter uid
	 * @return 	void	put PID from newsletter page in this->id
	 */
	function get_pid($newsletter_uid){
		global $TYPO3_DB;
		
		$query = $TYPO3_DB->exec_SELECTquery(
			'pid',
			'pages',
			'uid = '.$newsletter_uid
		);
		
		$parent = $TYPO3_DB->sql_fetch_assoc($query);
		$this->newsletter_uid = $newsletter_uid;
		$this->id = $parent['pid'];
	}
	
	/*
	 * get news uid to send
	 * 
	 * return	array		the news
	 */
	function get_ttnews_uid(){
		global $TYPO3_DB;
		
		$query = $TYPO3_DB->exec_SELECTquery(
			'*',
			'tt_news',
			'tx_dkdnews2directmail_senddmail = 1 '.
				'AND tx_dkdnews2directmail_dmail_uid = 0'.
				t3lib_BEfunc::BEenableFields('tt_news').
				t3lib_BEfunc::deleteClause('tt_news'),
			'',
			'tstamp ASC',
			'1'
		);
		return $TYPO3_DB->sql_fetch_assoc($query);
	}
}
?>
